rasa_core/policies/torch_policy.py

```diff
@@ -0,0 +1,234 @@
+                                                 )
+            self.input_size = input_size
+            self.output_size = output_size
+
+        val_split = kwargs.get("validation_split", 0.1)
+        train_size = int(shuffled_X.shape[0] * (1-val_split))
+
+        # convert to float since model requires float
+        train_X = torch.tensor(shuffled_X[:train_size]).type(torch.float)
+        train_y = torch.tensor(shuffled_y[:train_size]).type(torch.float)
+        val_X = torch.tensor(shuffled_X[train_size:]).type(torch.float)
+        val_y = torch.tensor(shuffled_y[train_size:]).type(torch.float)
+
+        del shuffled_X, shuffled_y
+
+        logger.info("Fitting model with {} total samples and validation split "
+                    "of {}".format(training_data.num_examples(), val_split))
+
+        # hyperparams and optimizer
+        epochs = kwargs.get("epochs", 30)
+        batch_size = kwargs.get("batch_size", 16)
+        optimizer = torch.optim.Adam(self.model.parameters(), lr=1e-3)
+        for ep in range(epochs):
+            permutation = torch.randperm(train_X.shape[0])
+            # for printing acc.
+            print_loss = 0
+            train_acc = 0
+            print_val_loss = 0
+            val_acc = 0
+            train_it = 0
+            val_it = 0
+
+            # train batches
+            for i in range(0, train_X.shape[0], batch_size):
+                optimizer.zero_grad()
+                idx = permutation[i:i + batch_size]
+                out = self.model(train_X[idx])[:, -1, :]
+                loss = -torch.mean(torch.sum(train_y[idx] * out, dim=1))
+                pred_y = torch.argmax(out, dim=1)
+                train_acc += accuracy_score(torch.argmax(train_y[idx], dim=1),
+                                            pred_y)
+                print_loss += loss.item()
+                train_it += 1
+                loss.backward()
+                optimizer.step()
+
+            for i in range(0, val_X.shape[0], batch_size):
+                val_it += 1
+                with torch.no_grad():
+                    out = self.model(val_X[i:i + batch_size])[:, -1, :]
+                    loss = -torch.mean(torch.sum(val_y[i:i + batch_size]
+                                                 * out, dim=1))
+                    pred_y = torch.argmax(out, dim=1)
+                    val_acc += accuracy_score(
+                        torch.argmax(val_y[i:i + batch_size], dim=1),
+                        pred_y)
+
+                    print_val_loss += loss.item()
+
+            logger.info("Epoch {} | Train loss/acc: {} / {} | "
+                        "Val loss/acc {} / {}"
+                        .format(ep + 1,
+                                round(print_loss, 3),
+                                round(train_acc / train_it, 3),
+                                round(print_val_loss, 3),
+                                round(val_acc / val_it, 3)))
+            self.current_epoch += 1
```

Comment:
`train()` can pass any number of epochs, but `self.current_epoch += 1` would only increment current epochs by 1

Reply:
Correct but this is to keep track of how many epochs we have trained on so that when we call `continue_training()` we know the number of epochs we have already trained on

Reply:
The number of epochs we trained on is based on `epochs = kwargs.get("epochs", 30)`. Should that be updated there?

Reply:
Yea we can just pass `self.current_epochs = epochs` but shouldn't `train()` always starts at 0 otherwise we call `continue_training()`?

Reply:
`continue_training` is executed only in interactive mode i.e., with `rasa_core.train --interactive`. It is possible to continue training in a non-interactive mode by loading a model from file with `rasa_core.train --core <persisted-model>`.

Reply:
I think `continue_training` function name is misleading.

Reply:
I see, that makes sense. But I think `self.current_epoch+=1` does the job of tracking number of epochs but just not as elegant as `self.current_epoch += epochs`. Is this what you were getting at?

Reply:
Yes